sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -135,9 +135,17 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
         PUSHED_FILTERS -> pushedDownFilters.mkString("[", ", ", "]"),
         INPUT_PATHS -> fsRelation.location.paths.mkString(", "))
 
+      // If the required attributes does not have the partitioning columns, we do not need
+      // to scan the partitioning columns. If partitioning columns are selected, the column order
+      // of partitionColumns is fixed in rdd. Thus, we always scan all the partitioning columns.
+      val scannedColumns = if (requiredAttributes.intersect(partitionSet).nonEmpty) {
+        readDataColumns ++ partitionColumns
+      } else {
+        readDataColumns
+      }
       val scan =
         DataSourceScanExec.create(
-          readDataColumns ++ partitionColumns,
+          scannedColumns,
```

Comment:
This is solution one that requires few code changes but gets most benefits. We can have another solution that covers more cases by changing the RDD generation.